### PR TITLE
Multi index container class

### PR DESCRIPTION
**Pull request by snschune**
@permcody @YaqiWang as discussed. tagging @dschwen 

Refs #14650

### CODE REVIEW

File: framework/include/utils/MultiIndex.h [hunk 1]
```diff
@@ -0,0 +1,496 @@
+/**********************************************************************/
+/*                     DO NOT MODIFY THIS HEADER                      */
+/* MAGPIE - Mesoscale Atomistic Glue Program for Integrated Execution */
```
Comment (permcody) [inline framework/include/utils/MultiIndex.h]: This header needs to be dropped.

**Review by permcody — CHANGES_REQUESTED** [review]
The rest of this looks fine. I suppose it should be if it's been in Magpie for awhile.

**Comment by snschune** [comment]
> The rest of this looks fine. I suppose it should be if it's been in Magpie for awhile.

Only the finest is good enough for the Magpie as @dschwen knows.

**Comment by snschune** [comment]
@permcody I am pretty sure that this failure is not related to this PR. I haven't even added documentation.

**Review by permcody — APPROVED** [review]

**Pull request merged**